src/main/java/cat/Cat.java

```diff
@@ -0,0 +1,67 @@
+package cat;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+
+import cat.command.Command;
+import cat.ui.Ui;
+
+/**
+ * The main class of the Duke program.
+ */
+public class Cat {
+    private static final String FILE_NAME = "duke.state";
+
+    private TaskList tasks;
+    private final Storage storage;
+    private final Ui ui;
+
+    /**
+     * The constructor for the Duke class.
+     *
+     * @param fileName file to store state to
+     */
+    public Cat(String fileName) {
+        assert fileName != null : "File name must not be null";
+
+        storage = new Storage(fileName);
+        ui = new Ui();
+        try {
+            tasks = storage.readTaskList();
+        } catch (IOException | ClassNotFoundException e) {
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Default constructor to launch from JavaFX.
+     */
+    public Cat() {
+        this(FILE_NAME);
+    }
+
+    public String getResponse(String input) {
+        try {
+            Command c = Parser.parse(input);
+            c.execute(tasks, ui, storage);
+        } catch (Parser.InvalidCommandType e) {
+            ui.showCommandNotFound(e.getCommand());
+        } catch (Parser.InvalidCommandData e) {
+            ui.showError(e);
+        }
+        return ui.flush();
+    }
+
+    /**
+     * Saves the state to storage.
+     */
+    public void save() {
+        try {
+            storage.writeTaskList(tasks);
+        } catch (FileNotFoundException e) {
+            System.out.println("Cannot find state file \"" + FILE_NAME + "\"");
+        } catch (IOException e) {
+            System.out.println("Cannot write to state file \"" + FILE_NAME + "\"");
+        }
+    }
```

Comment:
Great that you separated a save() function from main function. Could improve even more by putting save() in a storage class.